### PR TITLE
#891, Making local variables for nested forge methods calls when  null check is needed.

### PR DESCRIPTION
**Pull request by sjaakd**
Optimization and fixing an NPE in the generated code.

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/conversion/SimpleConversion.java [hunk 1]
```diff
@@ -39,7 +38,7 @@
     @Override
     public Assignment to(ConversionContext conversionContext) {
         String toExpression = getToExpression( conversionContext );
-        return AssignmentFactory.createTypeConversion(
+        return new TypeConversion(
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/conversion/SimpleConversion.java]: new line

File: processor/src/main/java/org/mapstruct/ap/internal/conversion/SimpleConversion.java [hunk 2]
```diff
@@ -49,7 +48,7 @@ public Assignment to(ConversionContext conversionContext) {
     @Override
     public Assignment from(ConversionContext conversionContext) {
         String fromExpression = getFromExpression( conversionContext );
-        return AssignmentFactory.createTypeConversion(
+        return new TypeConversion(
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/conversion/SimpleConversion.java]: new line.

File: processor/src/main/java/org/mapstruct/ap/internal/model/MappingBuilderContext.java [hunk 3]
```diff
@@ -81,21 +81,21 @@
...
-         * <li>Direct Assignment (empty TargetAssignment)</li>
+         * <li>SourceRHS Assignment (empty TargetAssignment)</li>
          * <li>null, no assignment found</li>
          * </ol>
          */
         @SuppressWarnings("checkstyle:parameternumber")
         Assignment getTargetAssignment(Method mappingMethod, String mappedElement, Type sourceType, Type targetType,
                                        String targetPropertyName, FormattingParameters formattingParameters,
-                                       SelectionParameters selectionParameters, String sourceReference,
+                                       SelectionParameters selectionParameters, SourceRHS sourceReference,
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/MappingBuilderContext.java]: old parameter name 'sourceReference`

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java [hunk 4]
```diff
@@ -399,7 +402,7 @@ private Assignment assignToPlainViaAdder(Type sourceType, Assignment rightHandSi
                 result = new AdderWrapper(
                     result,
                     method.getThrownTypes(),
-                    getSourceRef(),
+                    getSourceRHS().getSourceReference(),
                     sourceType
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: can be optimized... mouth source type and source reference are now present in Freemarker templates.

**Comment by agudian** [comment]
Looks alright to me (although I only understand half of it). Could you also rename the first commit? ("#891 Implementation and unit est")

**Comment by sjaakd** [comment]
> Looks alright to me (although I only understand half of it). Could you also rename the first commit? ("#891 Implementation and unit est")

No problem.. I'm worried though. I tried to refactor the `AdderWrapper` as well to follow similar lines. Did not manage yet. I can see were it goes wrong. I'll do a follow up on that one some day. 

This whole stuff is getting too fuzzy . Perhaps for 2.0 we should refactor this completely. Get rid of the wrapper mechanism and split all cases out in individual templates for each case. Its perhaps a lot of work, but makes it more transparent. 

I'll merge

**Pull request merged**